zyte_parsers/gtin.py

```diff
@@ -0,0 +1,161 @@
+import re
+from typing import Optional, Union
+
+import attr
+from gtin.validator import is_valid_GTIN
+from stdnum import isbn, ismn, issn
+
+from . import SelectorOrElement
+from .api import input_to_element
+from .utils import extract_text
+
+
+@attr.s(frozen=True, auto_attribs=True)
+class Gtin:
+    type: str
+    value: str
+
+
+GTIN_MATCH_SPECIAL_CHARACTER_REGEX = re.compile(r"[^0-9a-zA-Z]")
+GTIN_MATCH_NON_NUMERIC_REGEX = re.compile(r"[^0-9]")
+
+# Consider only those prefix which have some numeric values as these
+# values interfere with the gtin id extraction
+GTIN_PREFIX = [
+    "isbn13",
+    "isbn10",
+    "ean13",
+    "gtin8",
+    "gtin12",
+    "gtin13",
+    "gtin14",
+]
+GTIN_PREFIX_REGEX = re.compile("|".join(GTIN_PREFIX), re.IGNORECASE)
+GTIN_CENTER_REGEX = re.compile(r"^\D*|\D*$")
+
+
+def extract_gtin(node: Union[SelectorOrElement, str]) -> Optional[Gtin]:
+    """Extract a GTIN (Global Trade Item Number) from a node that contains its text.
+
+    It detects the GTIN type and returns it together with the cleaned GTIN
+    value. The following types are supported: `isbn10`, `isbn13`, `issn`,
+    `ismn`, `upc`, `gtin8`, `gtin13`, `gtin14`.
+
+    :param node: Node including the GTIN text.
+    :return: A GTIN item.
+    """
+    gtin: Optional[str]
+    if isinstance(node, str):
+        gtin = node
+    else:
+        node = input_to_element(node)
+        gtin = extract_text(node)
+    gtin_id = extract_gtin_id(gtin)
+    gtin_class = gtin_classification(gtin_id)
+    if gtin_class:
+        assert isinstance(gtin_id, str)
+        return Gtin(gtin_class, gtin_id)
+    return None
+
+
+def remove_gtin_numeric_prefix(gtin_code: str) -> str:
```

Comment:
Can we make this one private, or should it be a part of public API?

Reply:
Looks like this one is fine to make private.